python/max_flow.py

```diff
@@ -0,0 +1,172 @@
+        )
+
+        if flow_bottleneck == -1:
+            if delta < min_weight:
+                break
+            delta //= 2
+            continue
+
+        for i, e in enumerate(augmenting_path):
+            if isinstance(e, mgp.Edge):
+                if augmenting_path[i - 1] == e.from_vertex.id:
+                    edge_flows[e.id] = edge_flows.get(e.id, 0) + flow_bottleneck
+                elif augmenting_path[i - 1] == e.to_vertex.id:
+                    edge_flows[e.id] = edge_flows.get(e.id, 0) - flow_bottleneck
+                else:
+                    raise Exception("path is not ordered correctly")
+
+        print(augmenting_path)
+
+        # store path as mgp.Path to return
+        for i, elem in enumerate(augmenting_path):
+            if i == 0:
+                path = mgp.Path(context.graph.get_vertex_by_id(augmenting_path[0]))
+            else:
+                if isinstance(elem, mgp.Edge):
+                    path.expand(elem)
+
+        return_paths_and_flows.append((path, flow_bottleneck))
+
+    return [mgp.Record(path=path, flow=flow) for path, flow in return_paths_and_flows]
+
+
+def DFS_path_finding(
+    context: mgp.ProcCtx,
+    path: List,
+    current_v: mgp.Vertex,
+    end_v: mgp.Vertex,
+    edge_property: str,
+    delta: mgp.Number,
+    edge_flows: Dict,
+) -> mgp.Number:
+    """
+    Finds augmenting path for max_flow algorithm using recursive DFS
+    with minimum edge weight delta, as defined by capacity scaling.
+
+    :param path: list for storing path, elements are
+                 alternating mgp.VertexId and mgp.Edge
+    :param delta: lower bound for path flow
+    param edge_flows: dict containing existing flows of edges
+
+    :return: flow_bottleneck, smallest remaining capacity on the path,
+             -1 if no path to end_node is found
+    """
+
+    # instead of using residual edges, we check for in_edges with flow
+    for edge in chain(current_v.out_edges, current_v.in_edges):
+        if edge.from_vertex == current_v:
+            to_v = edge.to_vertex
+            remaining_capacity = edge.properties[edge_property] - edge_flows.get(
+                edge.id, 0
+            )
+        else:
+            to_v = edge.from_vertex
+            remaining_capacity = edge_flows.get(edge.id, 0)
+
+        if to_v.id in path:
+            continue
+
+        if remaining_capacity > delta:
+
+            path.append(edge)
+            path.append(to_v.id)
+
+            if to_v.id == end_v.id:
+                # found path
+                return remaining_capacity
+
+            flow_bottleneck = DFS_path_finding(
+                context, path, to_v, end_v, edge_property, delta, edge_flows
+            )
+            if flow_bottleneck != -1:
+                # function call found path, propagate back
+                return min(remaining_capacity, flow_bottleneck)
+
+    # no path found with this vertex, remove it and its edge
+    del path[-2:]
```

Comment:
Is there a better way to remove last two elements in Python?

Reply:
path = path[:-2] ? why is it better?